_posts/2018-08-14-grpc-with-protobuf.md

```diff
@@ -0,0 +1,225 @@
+
+Since we won't be generating any code, we'll need to write the message classes ourselves.  There are four methods, each which have a request and a response type.  This means we need to make eight messages:
+
+```java
+  static final class CreateRequest {
+    byte[] key;
+    byte[] value;
+  }
+
+  static final class CreateResponse {
+  }
+
+  static final class RetrieveRequest {
+    byte[] key;
+  }
+
+  static final class RetrieveResponse {
+    byte[] value;
+  }
+
+  static final class UpdateRequest {
+    byte[] key;
+    byte[] value;
+  }
+
+  static final class UpdateResponse {
+  }
+
+  static final class DeleteRequest {
+    byte[] key;
+  }
+
+  static final class DeleteResponse {
+  }
+```
+
+Because GSON uses reflection to determine how the fields in our classes map to the serialized JSON, we don't need to annotate the messages.
+
+Our client and server logic will use the request and response types, but gRPC needs to know how to produce and consume these messages.  To do this, we need to implement a [`Marshaller`](https://grpc.io/grpc-java/javadoc/io/grpc/MethodDescriptor.Marshaller.html).  A marshaller knows how to convert from an arbitrary type to an `InputStream`, which is then passed down into the gRPC core library.  It is also capable of doing the reverse transformation when decoding data from the network.  For GSON, here is what the marshaller looks like:
+
+```java
+  static <T> Marshaller<T> marshallerFor(Class<T> clz) {
+    Gson gson = new Gson();
+    return new Marshaller<T>() {
+      @Override
+      public InputStream stream(T value) {
+        return new ByteArrayInputStream(gson.toJson(value, clz).getBytes(StandardCharsets.UTF_8));
+      }
+
+      @Override
+      public T parse(InputStream stream) {
+        return gson.fromJson(new InputStreamReader(stream, StandardCharsets.UTF_8), clz);
+      }
+    };
+  }
+```
+
+Given a `Class` object for a some request or response, this function will produce a marshaller.  Using the marshallers, we can compose a full `MethodDescriptor` for each of the four CRUD methods.  Here is an example of the Method descriptor for _Create_:
+
+```java
+  static final MethodDescriptor<CreateRequest, CreateResponse> CREATE_METHOD =
+      MethodDescriptor.newBuilder(
+          marshallerFor(CreateRequest.class),
+          marshallerFor(CreateResponse.class))
+          .setFullMethodName(
+              MethodDescriptor.generateFullMethodName(SERVICE_NAME, "Create"))
+          .setType(MethodType.UNARY)
+          .build();
+```
+
+Note that if we were using Protobuf, we would use the existing Protobuf marshaller, and the method descriptors would be generated automatically.
```

Comment:
an inline link at  "existing protobuf marshaller" in code would be a good jump off if someone wants to cmpare.

Reply:
Done.